packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,161 @@
+import { StaticDataView } from '../../data-view';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView, env: Env): PreprocessorBucket {
+    const preprocessors: Preprocessor[] = [];
+
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      env,
+      preprocessors,
+    });
+  }
+
+  public readonly conditions: Map<string, Preprocessor>;
+  public readonly ineligible: Set<number>;
+
+  public env: Env;
+
+  constructor({ env, preprocessors = [] }: { env: Env; preprocessors?: Preprocessor[] }) {
+    this.env = env;
+
+    this.ineligible = new Set();
+    this.conditions = new Map();
+
+    this.update({ added: preprocessors });
+  }
+
+  private build() {
+    // Build filters to preprocessor set
+    // We'll use `Set` instead of `Array` to effectively remove duplicates
+    const bindings = new Map<number, Set<Preprocessor>>();
+
+    for (const preprocessor of this.conditions.values()) {
+      // Remove unused preprocessor
+      if (!(preprocessor.positives.size + preprocessor.negatives.size)) {
+        this.conditions.delete(preprocessor.condition);
+        continue;
+      }
+
+      for (const positive of preprocessor.positives) {
+        const binding = bindings.get(positive) ?? new Set();
+
+        binding.add(preprocessor);
+        bindings.set(positive, binding);
+      }
+
+      for (const negative of preprocessor.negatives) {
+        const binding = bindings.get(negative) ?? new Set();
+
+        binding.add(preprocessor);
+        bindings.set(negative, binding);
+      }
+    }
+
+    // Update ineligible based on bindings
+    this.ineligible.clear();
+
+    for (const [filter, preprocessors] of bindings.entries()) {
+      for (const one of preprocessors) {
+        if (!one.isEnvQualifiedFilter(this.env, filter)) {
+          this.ineligible.add(filter);
+          break;
+        }
+      }
+    }
+  }
+
+  public get isFilterEligible() {
+    return (filterId: number) => !this.ineligible.has(filterId);
+  }
```

Comment:
I would keep this as a normal function and make sure that when we call it is done with a lambda function to not have to worry about the binding.

Reply:
Looks nice to me!

Reply:
I'm continuing this at https://github.com/ghostery/adblocker/pull/3782#discussion_r1488534834